test/media-groups.test.js

```diff
@@ -0,0 +1,719 @@
+  assert.strictEqual(segmentLoaderPlaylist, media, 'set playlist on segment loader');
+  assert.strictEqual(mediaType.activePlaylistLoader, playlistLoader,
+    'set active playlist loader for media group');
+});
+
+QUnit.test('activeTrack returns the correct audio track', function(assert) {
+  const type = 'AUDIO';
+  const settings = { mediaTypes: MediaGroups.createMediaTypes() };
+  const tracks = settings.mediaTypes[type].tracks;
+  const activeTrack = MediaGroups.activeTrack[type](type, settings);
+
+  assert.equal(activeTrack(), null, 'returns null when empty track list');
+
+  tracks.track1 = { id: 'track1', enabled: false };
+  tracks.track2 = { id: 'track2', enabled: false };
+  tracks.track3 = { id: 'track3', enabled: false };
+
+  assert.equal(activeTrack(), null, 'returns null when no active tracks');
+
+  tracks.track3.enabled = true;
+
+  assert.strictEqual(activeTrack(), tracks.track3, 'returns active track');
+
+  tracks.track1.enabled = true;
+
+  // video.js treats the first enabled track in the track list as the active track
+  // so we want the same behavior here
+  assert.strictEqual(activeTrack(), tracks.track1, 'returns first active track');
+});
+
+QUnit.test('activeTrack returns the correct subtitle track', function(assert) {
+  const type = 'SUBTITLES';
+  const settings = { mediaTypes: MediaGroups.createMediaTypes() };
+  const tracks = settings.mediaTypes[type].tracks;
+  const activeTrack = MediaGroups.activeTrack[type](type, settings);
+
+  assert.equal(activeTrack(), null, 'returns null when empty track list');
+
+  tracks.track1 = { id: 'track1', mode: 'disabled' };
+  tracks.track2 = { id: 'track2', mode: 'hidden' };
+  tracks.track3 = { id: 'track3', mode: 'disabled' };
+
+  assert.equal(activeTrack(), null, 'returns null when no active tracks');
+
+  tracks.track3.mode = 'showing';
+
+  assert.strictEqual(activeTrack(), tracks.track3, 'returns active track');
+
+  tracks.track1.mode = 'showing';
+
+  // video.js treats the first enabled track in the track list as the active track
+  // so we want the same behavior here
+  assert.strictEqual(activeTrack(), tracks.track1, 'returns first active track');
+});
+
+QUnit.test('activeGroup returns the correct audio group', function(assert) {
+  const type = 'AUDIO';
+  let media = null;
+  const settings = {
+    mediaTypes: MediaGroups.createMediaTypes(),
+    masterPlaylistLoader: {
+      media: () => media
+    }
+  };
+  const groups = settings.mediaTypes[type].groups;
+  const tracks = settings.mediaTypes[type].tracks;
+  const activeTrack = MediaGroups.activeTrack[type](type, settings);
+  const activeGroup = MediaGroups.activeGroup(type, settings);
+
+  assert.equal(activeGroup(), null, 'returns null when no media in masterPlaylistLoader');
+
+  media = { attributes: { } };
+  groups.main = [{ id: 'en' }, { id: 'fr' }];
+
+  assert.strictEqual(activeGroup(), groups.main,
+    'defaults to main audio group when media does not specify audio group');
```

Comment:
well thats not really a scenario that would happen. the `main` group is only created if the master manifest doesn't specify any audio groups. We would run into this problem if a master manifest specifies audio groups, but then has a `STREAM-INF` that does not have an `AUDIO` attribute, which I suppose is possible because of the wording of the spec, though ive never seen a playlist like that. Maybe we want to rethink how it falls back in such an event. We could first try to fall back to the group with attribute `DEFAULT` and if there is none, fallback to some private `main_` group that we always create regardless?

Reply:
I think it might be overboard, good as is with that explanation 👍